File: asyncawait-ts/basics04.ts
import {sleep} from "./lib";

function main() {
    (async () => {
        await doWorld();
        console.log('Done');
    })();
    // cannot block
}

async function doWorld() {
    const p1 = (async () => {
        await sleep(2000);
        console.log('World 2');
    })();
    const p2 = (async () => {
        await sleep(1000);
        console.log('World 1');
    })();
    console.log('Hello');
    await Promise.all([p1, p2]);
}

main();

File: coroutine-py/basics04.py
import asyncio


def main():
    asyncio.run(async_main())


async def async_main():
    await do_world()
    print('Done')


async def do_world():
    async with asyncio.TaskGroup() as tg:
        tg.create_task(do_world_do_world_2())
        tg.create_task(do_world_do_world_1())
        print('Hello')


async def do_world_do_world_2():
    await asyncio.sleep(2)
    print('World 2')


async def do_world_do_world_1():
    await asyncio.sleep(1)
    print('World 1')


if __name__ == "__main__":
    main()
